presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java

```diff
@@ -16,53 +16,284 @@
+            Iterator<Symbol> iterator2 = o2.getPartitionBy().iterator();
+
+            while (iterator1.hasNext() && iterator2.hasNext()) {
+                Symbol symbol1 = iterator1.next();
+                Symbol symbol2 = iterator2.next();
+
+                int partitionByComparison = symbol1.compareTo(symbol2);
+                if (partitionByComparison != 0) {
+                    return partitionByComparison;
+                }
+            }
+
+            if (iterator1.hasNext()) {
+                return 1;
+            }
+            if (iterator2.hasNext()) {
+                return -1;
+            }
+            return 0;
+        }
+
+        private int compareOrderBy(WindowNode o1, WindowNode o2)
+        {
+            Iterator<Symbol> iterator1 = o1.getOrderBy().iterator();
+            Iterator<Symbol> iterator2 = o2.getOrderBy().iterator();
+
+            while (iterator1.hasNext() && iterator2.hasNext()) {
+                Symbol symbol1 = iterator1.next();
+                Symbol symbol2 = iterator2.next();
+
+                int orderByComparison = symbol1.compareTo(symbol2);
+                if (orderByComparison != 0) {
+                    return orderByComparison;
+                }
+                else {
+                    int sortOrderComparison = o1.getOrderings().get(symbol1).compareTo(o2.getOrderings().get(symbol2));
+                    if (sortOrderComparison != 0) {
+                        return sortOrderComparison;
+                    }
+                }
+            }
+
+            if (iterator1.hasNext()) {
+                return 1;
+            }
+            if (iterator2.hasNext()) {
+                return -1;
+            }
+            return 0;
+        }
+    }
+
+    /**
+     * Looks for the pattern (ProjectNode*)TargetNode, and rewrites it to TargetNode(ProjectNode*),
+     * returning an empty option if it can't rewrite the projects, for example because they rely on
+     * the output of the TargetNode.
+     * @param projects the nodes above the target, bottom first.
+     */
+    private static <N extends PlanNode> Optional<N> pullUnaryNodeAboveProjects(
+            Class<N> targetClass,
+            N target,
+            List<ProjectNode> projects)
+    {
+        if (projects.isEmpty()) {
+            return Optional.of(target);
         }
 
-        ImmutableMap.Builder<Symbol, WindowNode.Function> functionsBuilder = ImmutableMap.builder();
-        functionsBuilder.putAll(parent.getWindowFunctions());
-        functionsBuilder.putAll(child.getWindowFunctions());
-
-        return Optional.of(new WindowNode(
-                parent.getId(),
-                child.getSource(),
-                parent.getSpecification(),
-                functionsBuilder.build(),
-                parent.getHashSymbol(),
-                parent.getPrePartitionedInputs(),
-                parent.getPreSortedOrderPrefix()));
+        PlanNode targetChild = target.getSources().get(0);
```

Comment:
use `getOnlyElement`

Reply:
After changing to be specific to WindowNode, this code will disappear.